job-server/src/spark.jobserver/WebApi.scala

```diff
@@ -531,6 +534,36 @@ class WebApi(system: ActorSystem,
   override def timeoutRoute: Route =
     complete(500, errMap("Request timed out. Try using the /jobs/<jobID>, /jobs APIs to get status/results"))
 
+  def determineProxyUser(aConfig: Config,
+                         authInfo: AuthInfo,
+                         contextName: String): (String, Config) = {
+    if (config.getBoolean("shiro.authentication") && config.getBoolean("shiro.use-as-proxy-user")) {
+      //proxy-user-param is ignored and the authenticated user name is used
+      val config = aConfig.withValue(SparkJobUtils.SPARK_PROXY_USER_PARAM,
+        ConfigValueFactory.fromAnyRef(authInfo.toString))
+      (authInfo.toString + "_" + contextName, config)
+    } else {
+      (contextName, aConfig)
+    }
+  }
+
+  def removeProxyUserPrefix(authInfo: AuthInfo, contextNames: Seq[String]): Seq[String] = {
+    if (config.getBoolean("shiro.authentication") && config.getBoolean("shiro.use-as-proxy-user")) {
+      val userNamePrefix = authInfo.toString + "_"
+      contextNames map { cName =>
+        {
+          if (cName.startsWith(userNamePrefix)) {
```

Comment:
yes, but if _ is part of the user name then you can have collisions
In your tests if you add a context for another user called: presidentskroob_newUser the tests start to fail
```
diff --git a/job-server/test/spark.jobserver/auth/SJSAuthenticatorSpec.scala b/job-server/test/spark.jobserver/auth/SJSAuthenticatorSpec.scala
index 9277348..161cf1e 100644
--- a/job-server/test/spark.jobserver/auth/SJSAuthenticatorSpec.scala
+++ b/job-server/test/spark.jobserver/auth/SJSAuthenticatorSpec.scala
@@ -54,6 +54,7 @@ securityManager.cacheManager = $cacheManager
 root = secret, admin
 guest = guest, guest
 presidentskroob = 12345, president
+presidentskroob_newUser = 12345, president
 darkhelmet = ludicrousspeed, darklord, schwartz
 lonestarr = vespa, goodguy, schwartz
 
diff --git a/job-server/test/spark.jobserver/auth/WebApiWithAuthenticationSpec.scala b/job-server/test/spark.jobserver/auth/WebApiWithAuthenticationSpec.scala
index fd80e6f..eae3948 100644
--- a/job-server/test/spark.jobserver/auth/WebApiWithAuthenticationSpec.scala
+++ b/job-server/test/spark.jobserver/auth/WebApiWithAuthenticationSpec.scala
@@ -76,9 +76,13 @@ class WebApiWithAuthenticationSpec extends FunSpec with Matchers with BeforeAndA
   private val routesWithoutProxyUser = routesWithTimeout(false, "1 s")
 
   private val USER_NAME = "presidentskroob"
+  private val USER_NAME2 = "presidentskroob_newUser"
   // set to some valid user
   private val authorization = new Authorization(new BasicHttpCredentials(USER_NAME, "12345"))
   private val authorizationInvalidPassword = new Authorization(new BasicHttpCredentials(USER_NAME, "xxx"))
+
+  private val authorization2 = new Authorization(new BasicHttpCredentials(USER_NAME2, "12345"))
+  private val authorizationInvalidPassword2 = new Authorization(new BasicHttpCredentials(USER_NAME2, "xxx"))
   private val authorizationUnknownUser = new Authorization(new BasicHttpCredentials("whoami", "xxx"))
   private val dt = DateTime.parse("2013-05-29T00Z")
   private val jobInfo = JobInfo("foo-1", "context", JarInfo("demo", dt), "com.abc.meme", dt, None, None)
@@ -206,6 +210,10 @@ class WebApiWithAuthenticationSpec extends FunSpec with Matchers with BeforeAndA
       //clear
       addedContexts.clear
       //add a context that has the user name as a prefix
+      Post("/contexts/" + USER_NAME2 + "_c").withHeaders(authorization2) ~>
+        sealRoute(testRoutes) ~> check {
+          status should be(OK)
+        }
       Post("/contexts/" + USER_NAME + "_c").withHeaders(authorization) ~>
         sealRoute(testRoutes) ~> check {
           status should be(OK)
```